ocaml/typing/ctype.ml

```diff
@@ -5747,7 +5747,12 @@ let check_decl_immediate env decl imm =
     | Error _, Some ty -> check_type_immediate env ty imm
 
 let is_always_global env ty =
-  Result.is_ok (check_type_immediate env ty Always_on_64bits)
+  (* We snapshot to keep this pure; see the mode crossing test that mentions
+     snapshotting for an example. *)
+  let snap = Btype.snapshot () in
```

Comment:
You could probably guard doing the snapshoting by `!Clflags.principal || Env.has_local_constraints env` to avoid doing it when it doesn't matter.

Reply:
OK, I made the change you suggested (and tested it).  Will merge when the tests pass.